core/deployment/src/main/java/io/quarkus/deployment/util/JandexUtil.java

```diff
@@ -375,4 +380,219 @@ public ClassNotIndexedException(DotName dotName) {
+                typeSignature(exception, result);
+            }
+        }
+
+        return result.toString();
+    }
+
+    private static boolean hasGenericSignature(MethodInfo method) {
+        // JVMS 16, chapter 4.7.9.1. Signatures:
+        //
+        // Java compiler must emit ...
+        //
+        // A method signature for any method or constructor declaration which is either generic,
+        // or has a type variable or parameterized type as the return type or a formal parameter type,
+        // or has a type variable in a throws clause, or any combination thereof.
+
+        if (!method.typeParameters().isEmpty()) {
+            return true;
+        }
+
+        {
+            Type type = method.returnType();
+            if (type.kind() == Kind.TYPE_VARIABLE
+                    || type.kind() == Kind.UNRESOLVED_TYPE_VARIABLE
+                    || type.kind() == Kind.PARAMETERIZED_TYPE) {
+                return true;
+            }
+        }
+
+        for (Type type : method.parameters()) {
+            if (type.kind() == Kind.TYPE_VARIABLE
+                    || type.kind() == Kind.UNRESOLVED_TYPE_VARIABLE
+                    || type.kind() == Kind.PARAMETERIZED_TYPE) {
+                return true;
+            }
+        }
+
+        for (Type type : method.exceptions()) {
+            if (type.kind() == Kind.TYPE_VARIABLE
+                    || type.kind() == Kind.UNRESOLVED_TYPE_VARIABLE) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean hasThrowsSignature(MethodInfo method) {
+        // JVMS 16, chapter 4.7.9.1. Signatures:
+        //
+        // If the throws clause of a method or constructor declaration does not involve type variables,
+        // then a compiler may treat the declaration as having no throws clause for the purpose of
+        // emitting a method signature.
+
+        // also, no need to check if an exception type is of kind PARAMETERIZED_TYPE, because
+        //
+        // JLS 16, chapter 8.1.2. Generic Classes and Type Parameters:
+        //
+        // It is a compile-time error if a generic class is a direct or indirect subclass of Throwable.
+
+        for (Type type : method.exceptions()) {
+            if (type.kind() == Kind.TYPE_VARIABLE
+                    || type.kind() == Kind.UNRESOLVED_TYPE_VARIABLE) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static void typeParameter(TypeVariable typeParameter, StringBuilder result) {
+        result.append(typeParameter.identifier());
+
+        if (hasImplicitObjectBound(typeParameter)) {
+            result.append(':');
+        }
+        for (Type bound : typeParameter.bounds()) {
+            result.append(':');
+            typeSignature(bound, result);
+        }
+    }
+
+    private static boolean hasImplicitObjectBound(TypeVariable typeParameter) {
+        // TODO is there a better way? :-/
```

Comment:
@n1hility any help here? This is the only thing I wasn't able to reconstruct from Jandex's public API.